### ISSUE TITLE
Solved: Powershell Invoke-WebRequest Inconsistency - Unauthorized

### ISSUE THREAD

**Issue by matt-cylanda** [Ylianst/MeshCentral#3407]
TLDR, I couldn't get the agent or settings files to download to new systems automatically via the direct links in MeshCentral.

I finally figured out... some of the meshid string for the groups include a $ which in powershell scripting is the character to indicate the beginning of a variable name. If you use the ` escape character before each dollar sign, the files download without a problem.

Gonna leave the rest of the thread here for general info and scripting reference.

-------------------
Hey guys,

I've got a strange one. I'm guessing it has to do with the IP filtering we use on our instance, but for new client installs I set up a powershell script that reaches out to our remote server, downloads the agent, then installs locally via the command line.

In the past, because I couldn't get it working, I was downloading the agent separately, placing it in a web share, then referencing the url for downloading in the script. I decided to try out the same process, but reference the agent download directly from the remote server instead using the agent download link:

![image](https://user-images.githubusercontent.com/65490814/148280832-6f44e239-f0c7-4c7c-8c6b-5800244441f7.png)

Where I'm running into issues is with one client group the powershell download command works fine. But for other groups I get the below:

```
iwr : Unauthorized
At line:1 char:1
+ iwr -UseBasicParsing -OutF $output $uri -UseDefaultCredential -Sessio ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExcept  
   ion
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```
If I place the download url link in a browser on that machine, the download comes up fine. Below is the basic powershell script I use:

```
# Set SSL version for download compatibility
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
# x64 meshagent url & output variables + download command
$uri = "https://meshcentral.server.com/meshagents?id=4&meshid=LONGRANDOMCODEID&installflags=0"
$output = "c:\temp\meshagent.exe"
iwr -UseBasicParsing -OutF $output $uri
```

Any ideas about the inconsistency?

**Comment by Ylianst** [comment]
Hi. Yes, this agent download URL will only work if you are logged into the MeshCentral server. Since your powershell download is lot logged in, it will not work.

One alternative is to turn on agent invite codes. Add this line in the domain section of the config.json:

```
"agentInviteCodes": true,
```

Then, go in the device group and add an invite code... like 1234:

![image](https://user-images.githubusercontent.com/1319013/148306452-4113a6cd-2385-4dd7-8261-1cabf80a5535.png)

Then, you can go to the invitation URL (in the dialog box you set to code), type in the code:

![image](https://user-images.githubusercontent.com/1319013/148306507-93beb563-d018-4030-ae5b-ef8b9c69fa7a.png)

You can then see all the agents you can downloads, grab the URL and use that one in PowerShell.

![image](https://user-images.githubusercontent.com/1319013/148306602-33507a1a-0362-4507-85fd-8da2cde31e37.png)

Let me know if that works.

**Comment by matt-cylanda** [comment]
I tried this out, but it's still inconsistent. With the group I added the invite code to, powershell iwr still reported 'unauthorized' in the download request.

In case I didn't specify the use case, we're trying to smooth out our software rollout and automate the install of software with variables + software keys in our scripting

So in the script above, our roll out tool script for all our endpoints looks something like this:

```
#Client MeshAgentID
$MeshAgentID = LONGRANDOMCODEID

# x64 meshagent url & output variables + download command
$uri = "https://meshcentral.server.com/meshagents?id=4&meshid=$MeshAgentID&installflags=0"
$output = "c:\temp\meshagent.exe"

iwr -UseBasicParsing -OutF $output $uri
```

The really strange thing is in some client groups, on systems that I definitely haven't logged into the portal on, this works fine. But on others, I get the unauthorized response. 

Would it be possible to have the download script just download from the below, and then run the exe with a flag indicating the group it should install to? I saw a -nodeid flag in the command line options.
```https://meshcentral.server.com/meshagents?id=4```

**Comment by matt-cylanda** [comment]
Ok...

I figured it out. I decided to look over the method used for the linux install side of things. I saw the installation relies on the .msh file being present on the execution/install of the meshagent, so I altered my scripting to do the same:

```
# Set SSL version for download compatibility
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

# x64 meshagent url & output variables + download command
New-Item -ItemType Directory -Force -Path 'C:\Program Files\Mesh Agent\'
Set-Location -Path 'C:\Program Files\Mesh Agent\'

$settingsurl = "https://meshcentral.server.com/ meshsettings?id=$MeshID"
$settingsoutput = "C:\Program Files\Mesh Agent\MeshAgent.msh"
$installerurl = "https://meshcentral.server.com/meshagents?id=4"
$installeroutput = "C:\Program Files\Mesh Agent\MeshAgent.exe"

#Download Agent
IWR -UseBasicParsing -OutF $settingsoutput -Uri $settingsurl
IWR -UseBasicParsing -OutF $installeroutput -Uri $installerurl

# Install MeshAgent
.\MeshAgent -fullinstall
```

Would it be possible in a future version to have the group MeshID present somewhere in the groups to have as a one-time click and copy to clipboard?
